rust/kernel/print.rs

```diff
@@ -0,0 +1,775 @@
+    pub static CRIT: [u8; LENGTH] = generate(bindings::KERN_CRIT);
+    pub static ERR: [u8; LENGTH] = generate(bindings::KERN_ERR);
+    pub static WARNING: [u8; LENGTH] = generate(bindings::KERN_WARNING);
+    pub static NOTICE: [u8; LENGTH] = generate(bindings::KERN_NOTICE);
+    pub static INFO: [u8; LENGTH] = generate(bindings::KERN_INFO);
+    pub static DEBUG: [u8; LENGTH] = generate(bindings::KERN_DEBUG);
+
+    /// The length of the fixed format strings for the `CONT` "level".
+    pub const LENGTH_CONT: usize = 7;
+
+    /// Generates the fixed format string for the kernel's [`printk`] for
+    /// the `CONT` "level".
+    ///
+    /// [`printk`]: ../../../../include/linux/printk.h
+    const fn generate_cont() -> [u8; LENGTH_CONT] {
+        // Ensure the `KERN_CONT` macro is what we expect.
+        assert!(bindings::KERN_CONT[0] == b'\x01');
+        assert!(bindings::KERN_CONT[1] == b'c');
+        assert!(bindings::KERN_CONT[2] == b'\x00');
+
+        const SUFFIX: &[u8; LENGTH_CONT - LENGTH_PREFIX] = b"%.*s\0";
+        [
+            bindings::KERN_CONT[0],
+            bindings::KERN_CONT[1],
+            SUFFIX[0],
+            SUFFIX[1],
+            SUFFIX[2],
+            SUFFIX[3],
+            SUFFIX[4],
+        ]
+    }
+
+    // Generate the format string at compile-time for the `CONT` "level".
+    //
+    // This avoids the compiler generating the contents on the fly in the stack.
+    //
+    // Furthermore, `static` instead of `const` is used to share the string
+    // for all the kernel.
+    pub static CONT: [u8; LENGTH_CONT] = generate_cont();
+}
+
+/// Prints a message via the kernel's [`printk`].
+///
+/// Hidden because it is called from a public macro -- it should never be
+/// called by users.
+///
+/// # Safety
+///
+/// The format string must be one of the ones in [`format_strings`].
+///
+/// (This ensures that the format string matches the variadic arguments we pass
+/// to [`printk`]).
+///
+/// [`printk`]: ../../../../include/linux/printk.h
+#[doc(hidden)]
+pub unsafe fn call_printk(
+    format_string: &[u8; format_strings::LENGTH],
+    module_name: &[u8],
+    string: &[u8],
+) {
+    // `printk` does not seem to fail in any path.
+    bindings::printk(
+        format_string.as_ptr() as _,
+        module_name.len() as c_int,
+        module_name.as_ptr(),
+        string.len() as c_int,
+        string.as_ptr(),
+    );
+}
+
+/// Prints a message via the kernel's [`printk`] for the `CONT` "level".
+///
+/// Follows the interface and qualifiers of [`call_printk`] to make things
+/// easier in [`print_macro!`].
+///
+/// Hidden because it is called from a public macro -- it should never be
+/// called by users.
+///
+/// # Safety
+///
+/// See [`call_printk`]'s safety contract.
+///
+/// [`printk`]: ../../../../include/linux/printk.h
+#[doc(hidden)]
+pub unsafe fn call_printk_cont(
+    format_string: &[u8; format_strings::LENGTH_CONT],
```

Comment:
Is this value simply always `CONT`?

Reply:
Yeah, I kept it as a parameter to match `call_printk` so that we can use `$call_fn` in the macro without having to put a conditional there.